client/db/src/lib.rs

```diff
@@ -474,13 +478,112 @@ fn cache_header<Hash: std::cmp::Eq + std::hash::Hash, Header>(
 	}
 }
 
+struct PinnedBlockCacheEntry<Block: BlockT> {
+	ref_count: u32,
+	pub justifications: Option<Option<Justifications>>,
+	pub body: Option<Option<Vec<Block::Extrinsic>>>,
+}
+
+impl<Block: BlockT> Default for PinnedBlockCacheEntry<Block> {
+	fn default() -> Self {
+		Self { ref_count: 0, justifications: None, body: None }
+	}
+}
+
+impl<Block: BlockT> PinnedBlockCacheEntry<Block> {
+	pub fn decrease_ref(&mut self) {
+		self.ref_count = self.ref_count.saturating_sub(1);
+	}
+
+	pub fn increase_ref(&mut self) {
+		self.ref_count = self.ref_count.saturating_add(1);
+	}
+
+	pub fn has_no_references(&self) -> bool {
+		self.ref_count == 0
+	}
+}
+
+struct PinnedBlockCache<Block: BlockT> {
+	cache: RwLock<LruCache<Block::Hash, PinnedBlockCacheEntry<Block>>>,
+}
+
+impl<Block: BlockT> PinnedBlockCache<Block> {
+	pub fn new() -> Self {
+		Self { cache: LruCache::new(NonZeroUsize::new(PINNING_CACHE_SIZE).unwrap()).into() }
+	}
+
+	pub fn bump_ref(&self, hash: Block::Hash) {
+		let mut cache = self.cache.write();
+		if cache.len() == PINNING_CACHE_SIZE {
+			log::warn!(target: "db-pinning", "Maximum size of pinning cache reached. Removing items to make space. max_size = {}", PINNING_CACHE_SIZE);
```

Comment:
did you mean to say "Remove items to make space"? Otherwise where is "removing" part?

Reply:
Correct, I will adjust the log message. I think we should still print if the maximum size has been reached, since this indicates that we are hoarding too many notifications somewhere in the node.